docs/manual/src/udl/remote_ext_types.md

```diff
@@ -0,0 +1,142 @@
+#[uniffi::remote(Enum)]
+enum LogLevel {
+    Error,
+    Warn,
+    Info,
+    Debug,
+    "Trace",
+}
+```
+
+## UDL
+
+Wrap the definition with `[Remote]` attribute:
+
+```idl
+[Remote]
+enum LogLevel {
+    "Error",
+    "Warn",
+    "Info",
+    "Debug",
+    "Trace",
+};
+```
+
+# External Types
+
+"External types" refer to types defined in other crates that use UniFFI.
+This normally means types from other crates in your workspace.
+
+## Proc-macros
+
+Proc-macro-based code can use external types automatically, without any extra code.
+
+## UDL
+
+Suppose you depend on the `DemoDict` type from another UniFFIed crate in your workspace.
+You can reference this type by using the `[External]` attribute to wrap a typedef describing the concrete type.
+
+```idl
+[External]
+typedef dictionary One;
+
+// Now define our own dictionary which references the external type.
+dictionary ConsumingDict {
+  DemoDict demo_dict;
+  boolean another_bool;
+};
+```
+
+Supported values for the typedef type:
+
+* Enums: `enum`
+* Records: `record`, `dictionary` or `struct`
+* Objects: `object`, `impl` or `interface`
+* Traits: `trait`, `callback` or `trait_with_foreign`
+
+# Special cases for remote types
+
+There are a few cases where remote types require extra handling in addition to the rules above.
+
+## Remote + External types
+
+Types that are remote and external require a `use_remote_type!` macro call.
+
+If `crate_a` defines [IpAddr](https://doc.rust-lang.org/std/net/enum.IpAddr.html) as a remote type, then `crate_b` can use that type with the following Rust code:
+
+```rust
+uniffi::use_remote_type!(IpAddr, crate_a);
+```
+
+## UDL
+
+UDL-users will also need to add the external type definition:
+
+```idl
+[External]
+typedef enum IpAddr;
+```
+
+## Remote custom types
+
+Types that are remote and custom require using the `remote` attribute with the `custom_type!` macro.
+
+```rust
+
+uniffi::custom_type!(StructFromOtherCrate, String, {
+    remote,
+    lower: |s| s.to_string(),
+    try_lift: |s| StructFromOtherCrate::try_from_string(s),
+});
+```
+
+## Foreign bindings
+
+The foreign bindings will also need to know how to access the external type,
+which varies slightly for each language:
+
+### Kotlin
+
+For Kotlin, "library mode" generation with `generate --library [path-to-cdylib]` is recommended when using external types.
+If you use `generate [udl-path]` then the generated code needs to know how to import
+the external types from the Kotlin module that corresponds to the Rust crate.
+By default, UniFFI assumes that the Kotlin module name matches the Rust crate name, but this can be configured in `uniffi.toml` with an entry like this:
+
+```
+[bindings.kotlin.external_packages]
+# Map the crate names from [External={name}] into Kotlin package names
+rust-crate-name = "kotlin.package.name"
+```
+
+### Swift
+
+For Swift, you must compile all generated `.swift` files together in a single
```

Comment:
I definitely agree, but I'm actually not sure what to advise a user.  I know how we build our components, but I don't exactly want to recommend that to others.  I think there's a `cargo swift` command that helps out with this, but don't know enough about that to recommend it.

The tutorial says "Note that these commands could be integrated as part of your gradle/Xcode build process", which seems like the best advice we have to offer.  I'd love to improve this story, are you interested in writing something up?

Reply:
Ok so my point is we need to guide user how to generate the swift bindings into a single module map. 

Uniffi does this properly already for bindgen tests, by combining the modulemaps here:

https://github.com/mozilla/uniffi-rs/blob/de6eef9dba8e4a463f0706edfa88649e71541aef/uniffi_bindgen/src/bindings/swift/test.rs#L164

So I just think we need to surface that to the uniffi bindgen cli tool? At least optin by some flag 

I see that 2 months ago you did some work on Swift bindgen, https://github.com/mozilla/uniffi-rs/blob/de6eef9dba8e4a463f0706edfa88649e71541aef/docs/manual/src/swift/uniffi-bindgen-swift.md?plain=1#L8

Perhaps you have already added support for merging of the generates swift bindings for multi crate setup?

Otherwise I think UniFFI Swift users will be hit with this issue: 

https://github.com/mozilla/uniffi-rs/pull/2087#issuecomment-2128711332

Reply:
Yeah, so I made the `uniffi-bindgen-swift` CLI which has a way to create a combined modulemap and also a flag for XCFramework compatibility.  It also allows generating only one kind of source file (.modulemap, .h, or .swift) at a time, so you can put different types in different directories.  It's very flexible but I'm not sure what to recommend for concrete use cases: how should users integrate it with an XCode project?  How should they create a SPM package?  Are there other common build systems?

I'm not a Swift dev.  I don't even know if I'm asking the right questions here.  I would love it if someone who does know Swift would write up some docs on this.  I'd be happy to help by describing what `uniffi-bindgen-swift` can do and adding whatever new features would simplify this process.

Reply:
The combined module's would be very useful for me, because then I can split up my current single rust library to multiple ones in a workspace for better compile times: https://github.com/mozilla/uniffi-rs/issues/2257#issuecomment-2422716680.

I don't need to generate SPM package for the one I reference above (Cove) because its an app not a library.

But creating an SPM, I have a different library written in rust, that I do it for, the SPM zip is created in GitHub actions: https://github.com/bitcoinppl/bbqr-swift/blob/master/.github/workflows/publish-spm.yml